File: src/rb/PhantasyIslandRubyRemoteControl/radio/radio_manager.rb
require 'socket.io-client-simple'
require 'thread'
require 'weakref'

require_relative 'wait_token'
require_relative 'api_debug'
require_relative 'api_scene'
require_relative 'api_fly'
require_relative 'api_radio'

class RadioManager
  attr_accessor :socket, :namespace, :scene_is_init
  attr_accessor :debugApi, :sceneApi, :flyApi, :radioApi

  def initialize
    @namespace = '/UserSide'
    @scene_is_init = false
    @pending_waiters = {} # wait_cmd -> Array of WeakRef
    @waiters_lock = Mutex.new

    # 初始化 Sub-API 模块
    @debugApi = DebugApi.new(self)
    @sceneApi = SceneApi.new(self)
    @flyApi = FlyApi.new(self)
    @radioApi = RadioApi.new(self)
  end

  def create_msg_timestamp_id
    (Time.now.to_f * 1000 * 100).to_i
  end

  def connect(url = 'http://127.0.0.1:60002', namespace = '/UserSide')
    @namespace = namespace
    reset
    
    # 使用 socket.io-client-simple 库
    # 注意：Ruby 的 socket.io 客户端库可能需要根据实际可用库进行调整
    @socket = SocketIO::Client::Simple.connect(url)
    _init_listener
  end

  def reset
    if @socket && @socket.connected?
      @socket.disconnect
    end
    @scene_is_init = false
  end

  def _init_listener
    rm = self
    @socket.on :connect do
      puts '[RadioManager] connected'
      rm._check_scene_status
    end

    @socket.on :disconnect do
      puts '[RadioManager] disconnected'
      rm.scene_is_init = false
    end

    @socket.on :message do |data|
      # 假设 data 是解析后的 Hash。如果不是，可能需要 JSON.parse(data)
      puts "[RadioManager] message: #{data}"
      rm.msg_dispatch(data)
    end

    @socket.on :error do |err|
      puts "[RadioManager] error: #{err}"
    end
  end

  def _check_scene_status
    _send('ping')
    _send('scene.getInitState')
  end

  def ping
    _send_and_wait_sync('ping', wait_cmd: 'pong')
  end

  def _send(cmd, data = nil)
    msg = { 'cmd' => cmd }
    msg.merge!(data) if data
    # 在 socket.io-client-simple 中，通常使用 emit
    @socket.emit :message, msg
  end

  # ---- 请求-响应: 核心 ----

  def _send_with_token(cmd, data = nil, wait_cmd: nil, post_processor: nil)
    wait_cmd ||= cmd
    time_base_id = create_msg_timestamp_id
    token = WaitToken.new(wait_cmd, time_base_id)
    token.set_post_processor(post_processor) if post_processor

    # 注册弱引用
    ref = WeakRef.new(token)
    @waiters_lock.synchronize do
      @pending_waiters[wait_cmd] ||= []
      @pending_waiters[wait_cmd] << ref
    end

    msg = { 'timestampIdPython' => time_base_id }
    msg.merge!(data) if data
    # 移除 nil 字段
    msg.delete_if { |k, v| v.nil? }

    _send(cmd, msg)
    token
  end

  # ---- 请求-响应: 便捷方法 ----

  def _send_and_wait_sync(cmd, data = nil, wait_cmd = nil, timeout = 3.0, post_processor: nil)
    token = _send_with_token(cmd, data, wait_cmd: wait_cmd, post_processor: post_processor)
    token.wait(timeout)
  end

  def _send_and_wait_token(cmd, data = nil, wait_cmd = nil, post_processor: nil)
    _send_with_token(cmd, data, wait_cmd: wait_cmd, post_processor: post_processor)
  end

  # Ruby 中简单的异步等待可以用线程模拟，或者直接返回 token
  def _send_and_wait_async(cmd, data = nil, wait_cmd = nil, timeout = 3.0, post_processor: nil)
    _send_with_token(cmd, data, wait_cmd: wait_cmd, post_processor: post_processor)
  end

  # ---- 弱引用分发 ----

  def _notify_waiters(cmd, data)
    timestamp_id = data['timestampIdPython']
    return false if timestamp_id.nil?

    @waiters_lock.synchronize do
      refs = @pending_waiters[cmd]
      return false if refs.nil? || refs.empty?

      surviving = []
      matched = false
      refs.each do |ref|
        begin
          token = ref.__getobj__
          if !matched && token.time_base_id == timestamp_id
            token.complete(data)
            matched = true
          else
            surviving << ref
          end
        rescue WeakRef::RefError
          # 已被回收
          next
        end
      end

      if surviving.empty?
        @pending_waiters.delete(cmd)
      else
        @pending_waiters[cmd] = surviving
      end
      matched
    end
  end

  def msg_dispatch(data)
    cmd = data['cmd']
    return if cmd.nil?

    if _notify_waiters(cmd, data)
      return
    end

    case cmd
    when 'pong'
      # ignore
    when 'sceneReset', 'sceneNotInit'
      _on_scene_reset(data)
    when 'sceneInit', 'sceneIsInit'
      _on_scene_init(data)
    else
      puts "[RadioManager] unknown cmd: #{cmd}, data: #{data}"
    end
  end

  def _on_scene_reset(data)
    puts "[RadioManager] handle sceneReset: #{data}"
    @scene_is_init = false
  end

  def _on_scene_init(data)
    puts "[RadioManager] handle sceneInit: #{data}"
    @scene_is_init = true
  end
end

File: src/php/PhantasyIslandPhpRemoteControl/radio/RadioManager.php
<?php

namespace PhantasyIslandPhpRemoteControl\Radio;

/**
 * RadioManager 的 PHP 实现。
 * 注：PHP 标准环境不支持异步，Socket.IO 通信通常使用第三方库如 Elephant.io。
 * 这里提供接口和逻辑骨架。
 */
class RadioManager
{
    public $socket = null; // 假设使用的 Socket.IO 客户端实例
    public $namespace = '/UserSide';
    public $scene_is_init = false;

    /** @var DebugApi */
    public $debugApi;
    /** @var SceneApi */
    public $sceneApi;
    /** @var FlyApi */
    public $flyApi;
    /** @var RadioApi */
    public $radioApi;

    // wait_cmd -> [WaitToken, ...]
    private $pending_waiters = [];

    public function __construct()
    {
        // 初始化 Sub-API 模块
        $this->debugApi = new DebugApi($this);
        $this->sceneApi = new SceneApi($this);
        $this->flyApi = new FlyApi($this);
        $this->radioApi = new RadioApi($this);
    }

    public function create_msg_timestamp_id()
    {
        return floor(microtime(true) * 1000 * 100);
    }

    public function connect($url = 'http://127.0.0.1:60002', $namespace = '/UserSide')
    {
        $this->namespace = $namespace;
        // 在此处初始化 Socket.IO 连接并设置回调
        // $this->socket = new Client($url);
        // $this->socket->on('message', [$this, 'msg_dispatch']);
        // $this->socket->connect();
        $this->check_scene_status();
    }

    private function check_scene_status()
    {
        $this->_send('ping');
        $this->_send('scene.getInitState');
    }

    public function ping()
    {
        return $this->_send_and_wait_sync('ping', null, 'pong');
    }

    /**
     * 底层发送
     */
    public function _send($cmd, $data = null)
    {
        $msg = ['cmd' => $cmd];
        if ($data) {
            $msg = array_merge($msg, $data);
        }
        // $this->socket->emit('message', $msg, $this->namespace);
    }

    /**
     * 发送并返回令牌
     */
    public function _send_with_token($cmd, $data = null, $wait_cmd = null, $post_processor = null)
    {
        if ($wait_cmd === null) {
            $wait_cmd = $cmd;
        }

        $time_base_id = $this->create_msg_timestamp_id();
        $token = new WaitToken($wait_cmd, $time_base_id);
        if ($post_processor !== null) {
            $token->setPostProcessor($post_processor);
        }

        // 注册等待者
        $this->pending_waiters[$wait_cmd][] = $token;

        $msg = ['timestampIdPython' => $time_base_id];
        if ($data) {
            foreach ($data as $k => $v) {
                if ($v !== null) $msg[$k] = $v;
            }
        }

        $this->_send($cmd, $msg);
        return $token;
    }

    public function _send_and_wait_sync($cmd, $data = null, $wait_cmd = null, $timeout = 3.0, $post_processor = null)
    {
        $token = $this->_send_with_token($cmd, $data, $wait_cmd, $post_processor);
        return $token->wait($timeout);
    }

    public function _send_and_wait_token($cmd, $data = null, $wait_cmd = null, $post_processor = null)
    {
        return $this->_send_with_token($cmd, $data, $wait_cmd, $post_processor);
    }

    /**
     * 消息分发，应由 Socket.IO 监听器调用。
     */
    public function msg_dispatch($data)
    {
        $cmd = $data['cmd'] ?? null;
        if ($this->notify_waiters($cmd, $data)) {
            return;
        }

        switch ($cmd) {
            case 'pong': break;
            case 'sceneReset':
            case 'sceneNotInit':
                $this->scene_is_init = false;
                break;
            case 'sceneInit':
            case 'sceneIsInit':
                $this->scene_is_init = true;
                break;
        }
    }

    private function notify_waiters($cmd, $data)
    {
        $timestamp_id = $data['timestampIdPython'] ?? null;
        if (!isset($this->pending_waiters[$cmd])) {
            return false;
        }

        $matched = false;
        $surviving = [];
        foreach ($this->pending_waiters[$cmd] as $token) {
            if (!$matched && $timestamp_id !== null && $token->time_base_id == $timestamp_id) {
                $token->complete($data);
                $matched = true;
            } else {
                $surviving[] = $token;
            }
        }

        if (empty($surviving)) {
            unset($this->pending_waiters[$cmd]);
        } else {
            $this->pending_waiters[$cmd] = $surviving;
        }

        return $matched;
    }
}
